### PR TITLE
refactor: Link landscapes directly to shared files, links and visualizations

### PR DESCRIPTION
**Pull request by josebui**
## Description

 Link landscapes directly to shared files, links and visualizations
 
### Checklist
- [x] Corresponding issue has been opened
- [x] New tests added

### Related Issues
- https://github.com/techmatters/terraso-web-client/issues/1228

### Related PRs
- https://github.com/techmatters/terraso-web-client/pull/1229
- https://github.com/techmatters/terraso-client-shared/pull/133

### CODE REVIEW

File: terraso_backend/apps/shared_data/views.py [hunk 1]
```diff
@@ -36,40 +38,74 @@
...
 
 class DataEntryFileUploadView(AuthenticationRequiredMixin, FormView):
+    @transaction.atomic
     def post(self, request, **kwargs):
         form_data = request.POST.copy()
         form_data["created_by"] = str(request.user.id)
         form_data["entry_type"] = DataEntry.ENTRY_TYPE_FILE
+        target_type = form_data.pop("target_type")[0]
+        target_slug = form_data.pop("target_slug")[0]
+        if target_type not in ["group", "landscape"]:
```
Comment (paulschreiber) [inline terraso_backend/apps/shared_data/views.py:48]: suggestion: make a `VALID_TARGET_TYPES` constant.

File: terraso_backend/apps/graphql/schema/data_entries.py [hunk 2]
```diff
@@ -77,32 +123,45 @@ class DataEntryAddMutation(BaseWriteMutation):
...
+            )
+
+        content_type = ContentType.objects.get(app_label="core", model=target_type)
+        model_class = content_type.model_class()
 
         try:
-            group = Group.objects.get(slug=group_slug)
-        except Group.DoesNotExist:
+            target = model_class.objects.get(slug=target_slug)
+        except Exception:
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/data_entries.py:154]: question: can you specify a more specific exception type?

**Review by david-code — APPROVED** [review]
Looks good to me

File: terraso_backend/apps/shared_data/permission_rules.py [hunk 3]
```diff
@@ -15,6 +15,40 @@
...
+        return user.memberships.approved_only().filter(group=target.get_default_group()).exists()
+    return False
+
+
+def is_user_allowed_to_view_data_entry(data_entry, user):
+    shared_resources = data_entry.shared_resources.all()
+    for shared_resource in shared_resources:
+        if is_target_member(user, shared_resource.target):
+            return True
+    return False
```
[the remark this answers is not in the document]
Comment (josebui) [reply terraso_backend/apps/shared_data/permission_rules.py:42]: @paulschreiber Yes, that can happen, so if the user has access to at least one of the targets where the resource is shared they can view it, that is why the algorithm exists early inside the for loop if the user is a member of any target.

File: terraso_backend/apps/graphql/schema/shared_resources.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        model = SharedResource
+        fields = ["id"]
+        interfaces = (relay.Node,)
+        connection_class = TerrasoConnection
+
+    def resolve_source(self, info, **kwargs):
+        return self.source
+
+    def resolve_target(self, info, **kwargs):
+        return self.target
```
[the remark this answers is not in the document]
Comment (josebui) [reply terraso_backend/apps/graphql/schema/shared_resources.py:51]: @david-code Yes, this has to do with the generic relationship, graphene is not able to do this automatically so it expects this to be defined explicitely

**Review by paulschreiber — APPROVED** [review]

**Pull request merged**